Feature/9

teabagk7 opened the pull request:
May be problem with "^" in string №40.
Your criticism and suggestions, pls.

teabagk7 committed:
git-bug origin

git-bug created.
added interaction with features: /2 and /3.

teabagk7 committed:
Update git-bug

teabagk7 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Список открытых issues ('git bug')](https://trello.com/c/WTTUFY3k/9-issues-git-bug)

akhismat commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+				summary=`git cat-file -p ${hash_of_summary:12:40}`
+				echo "$hash_of_commit" "$summary"
+			else
+				if [ "$1" = "--resolved" ]
```
Давайте использовать здесь switch вместо вложенных if-else

teabagk7 replied:
done

akhismat commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+				then	
+					hash_of_summary=`git ls-tree $hash_of_commit summary`
+					summary=`git cat-file -p ${hash_of_summary:12:40}`
+					echo "$hash_of_commit" "$summary"
```
Эти 3 строчки используются 6 раз в этом файле. Пожалуйста, вынесите это в отдельную функцию.

teabagk7 replied:
done

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+if [ "$1" = "new" ]
+then
+	git bug-new $2 $3
+else
```
зачем тут `else`, лучше в его if'е просто поставить `exit`, ну или то, что @AnnaKhismatullina предложила

teabagk7 replied:
done

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
-echo Sample for git bug
+if [ "$1" = "new" ]
+then
+	git bug-new $2 $3
```
ну вообще лучше все параметры передавать через `"$@"`. Если нужно при этом "выбросить" первый `$1`, то использовать команду `shift`, которая сдвинет аргументы (второй сделает первым, третий - вторым и т.д.)

teabagk7 replied:
done

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+	then
+		git bug-resolve $2
+	else
+		temp=`git show-ref --hash ISSUES`
```
имя `temp` не говорящее - нужно называть так, чтобы было понятно по коду, что это и для чего

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+	else
+		temp=`git show-ref --hash ISSUES`
+		PARENT=`git log ISSUES --pretty=format:"%P"`
+		until [ "$PARENT" = "" ]
```
может традиционный `while`? )

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+		until [ "$PARENT" = "" ]
+		do
+			hash_of_blob=`git cat-file -p $temp^{tree}`
+			hash_of_commit=`git cat-file -p ${hash_of_blob:12:40}`
```
а что тут `:12:40` обозначает? вообще, в нетривиальных местах кода не запрещается делать комменты )

teabagk7 replied:
выделяем подстроку(hash) из строки

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+					fi
+				fi
+			fi
+			temp=$temp^
```
а что здесь хотелось от `^`? если просто в `git log` следующий передать, ну так там и прописать

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+		if [ "$1" = "-a" ]
+		then
+			hash_of_summary=`git ls-tree $hash_of_commit summary`
+			summary=`git cat-file -p ${hash_of_summary:12:40}`
```
лучше в двойные кавычки заключить:
`summary="`git cat-file ...`"`
возможно, в других местах тоже, не присматривался

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,72 @@
+else
+	if [ "$1" = "resolve" ]
+	then
+		git bug-resolve $2
```
аналогично все переменные передавать

teabagk7 committed:
Update git-bug

akhismat commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+}
+
+
+if [ "$1" = "new" ]
```
Давай объявим переменную, назовем ее типа command или subcommand и присвоем ей значение параметра $1, будет четь понятнее, что происходит:)

akhismat commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+
+if [ "$1" = "new" ]
+then
+        shift
```
shift, пожалуй, можно делать всегда, не стоит плодить одинаковые строки в каждом if'e

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
-#!/bin/sh
-echo Sample for git bug
+#!/bin/bash
+print_issue()
```
Строку `#!...` лучше отделить пустой строкой от остального кода.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+        echo "$hash_of_issue" "$summary"
+}
+
+
```
Тут целых две пустые строки, а везде ниже - ни одной. Предлагаю более равномерно разделять код на логические части одной пустой строкой.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+        git bug-resolve "$@"
+        exit 0
+fi
+if  git show-ref --quiet --verify ISSUES;
```
Точка с запятой в конце, кажется, не нужна? Вообще, я бы весь код отсюда и до конца вынес в отдельную функцию с говорящим названием вроде _list_issues()_

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+then 
+        exit 0
+fi
+cont_of_issues="`git ls-tree ISSUES`"
```
`count_of_issues`?

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+cont_of_issues="`git ls-tree ISSUES`"
+while [ ! "$cont_of_issues" = "" ]
+do      
+        string_length=94 #ТАК НЕОБХОДИМО потому что не работает  string_length="`expr index "$cont_of_issues" \n`" КАК НАЙТИ КОНЕЦ СТРОКИ?
```
а что за конец строки нужно найти? есть же команда _wc_, есть вся мощь _sed_ и _awk_

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+             #string_length="`expr "$cont_of_issues" : '.*'`"
+        #fi
+        string=${cont_of_issues:0:$string_length}
+        hash_of_issue="${string:12:40}"
```
хотелось бы без хардкода магических цифр обойтись. На крайний случай - в отдельные говорящие константы вынести.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,62 @@
+	                fi
+	                ;;
+        esac
+		if [ "$#" -eq 0 ]
```
отступ от края непонятный. Видимо, табы вместо пробелов тут? Все отступы в коде нужно делать одинаковым образом - например пробелами.

Tatiotaeryy committed:
Tabulation has been restored.

- add tabulation
- now print right hash of issue
- exit if there are no issues.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,58 @@
+then
+	shift
+	git bug-new "$@"
+	exit 0
```
кажется, лучше написать `exit $?` - выходить с кодом возврата `git bug-new`. Только надо проверить, что это правильно работает.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,58 @@
+
+if ! git show-ref --quiet --verify refs/heads/ISSUES
+then 
+	exit 0
```
а что данный выход означает? нет ишуёв?

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,58 @@
+	exit 0
+fi
+
+cont_of_issues="`git ls-tree ISSUES`"
```
`count_of_issues`? или тут какое-то секретное сокращение? ранее уже спрашивалось - don't ignore замечания, сделанные ранее

teabagk7 replied:
content

teabagk7 committed:
HUMANITY IS OVERRATED

Унифицировано для всех систем, скриптов.
Написали usage(минимальный).
Привели код к понятному виду.
Начали пользоваться awk.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+
+list_issues()
+{
+	
```
лишняя пустая строка

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+list_issues()
+{
+	
+	if ! git show-ref --quiet --verify $REF
```
давайте в одну строчку тоже превратим с помощью конструкции `[ ... ] && return`

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+	local issues_names="$(git ls-tree --name-only $REF)"
+	while [ ! "$issues_names" = "" ]
+	do      
+		
```
лишняя пустая строка

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+	while [ ! "$issues_names" = "" ]
+	do      
+		
+		local NAME_LENGTH=40
```
эта переменная нигде не используется

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+			"R")
+				[ "$status" = "resolved" ] && print_issue $issue
+				;;
+
```
лишняя пустая строка

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+	echo new $'\t\t\t' create new issue
+	echo resolve $'\t\t' resolve issue
+	echo
+	
```
лишняя пустая строка

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+
+usage()
+{
+	echo "usage: git bug [-a | --open | --resolved] <subcommand>"
```
1) добавить опцию _-h_
2) без параметров `git bug` мы же тоже умеем и хотим запускать? как тогда скрипт отработает?
3) судя по скрипту, тут `<subcommand>` тоже опционален, т.е.: `git bug [-h | -a | --open | --resolved | <subcommand>]`

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+
+main()
+{
+	
```
лишняя пустая строка. Мы либо везде принимаем за правило отступать от начала скобки, либо нигде.

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+	shift
+	
+	case "$subcommand" in
+		"-h")
```
давайте ещё тут же `"--help"` добавим (по-моему, просто через `|` можно прямо тут прописать)

teabagk7 replied:
--help должен по умолчанию вызвать man,  насколько мне известно

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+	fi
+
+	source "$workingdir/git-bug-$subcommand"
+	if [ ! type "git_bug_$subcommand" ]
```
а это функция такая должна появиться после предыдущего `source`? если да, то мы, получается, завязываемся на детали реализации того скрипта. Почему бы здесь просто сразу не выполнить скрипт `"$workingdir/git-bug-$subcommand" "$@"` и вернуть его код возврата?

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,117 @@
+
+		esac
+
+		local issues_names="$(echo "$issues_names" | awk ' NR != 1 {print $0}')"
```
это будет работать, но на большом количестве ишуёв устроенный таким образом наш цикл `while` будет медленным. Ты, получается, на каждый проход по циклу перезаписываешь в новый массив всё, кроме первой строки. Стоит подумать об оптимизации - возможно, в отдельной задаче, чтобы сейчас не сильно тратить на это время.

teabagk7 committed:
WE DID EVERYTHING WE COULD HAVE DONE

убрали пустые строчки
добавили -h
пытались оптимизировать цикл while
вызываем подкоманду напрямую

teabagk7 committed:
HOT FIX

забыли про проверку, вернули

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,99 @@
+				;;
+		esac
+
+		let "issue_number = issue_number + 1"
```
`let` - встроенная команда _bash_, но не _sh_. Первой строчкой в скрипте же указан _sh_, значит плюшки баша работать не будут. Давай тут просто напишем стандартный sh-инкремент:

```
issue_number=$((issue_number+1))
```

А больше нигде тут на баш завязки нет? я, вроде, заметил только это место.

teabagk7 replied:
Похоже, что только здесь.
Поправлю

kevromster commented on src/git-bug:
```diff
@@ -1,2 +1,99 @@
+		usage
+		exit 1
+	fi
+	source $workingdir/git-bug-$subcommand "$@"
```
Тут не нужен `source`, зачем? Он тебе фактически "вставит" содержимое другого скрипта в это место, а это не то, что мы хотим. Просто запускаем другой скрипт без всяких `source`, лучше с кавычками на случай пробелов в путях:

```
"$workingdir/git-bug-$subcommand" "$@"
```

В остальном скрипт выглядит ок.

kevromster reviewed (changes requested):
Два замечания, и можно будет мерджить.

teabagk7 committed:
UPDATE TO MERGE

Удалили завязку на bash
убрали source

kevromster reviewed (approved):
LGTM
